src/coreclr/jit/codegenlinear.cpp

```diff
@@ -758,12 +749,24 @@ void CodeGen::genCodeForBBlist()
                 //
                 // During emitter, this information will be used to calculate the loop size.
                 // Depending on the loop size, decision of whether to align a loop or not will be taken.
+                //
+                // In the emitter, we need to calculate the loop size from `block->bbJumpDest` through
+                // `block` (inclusive). Thus, we need to ensure there is a label on the lexical fall-through
+                // block, even if one is not otherwise needed, to be able to calculate the size of this
+                // loop (loop size is calculated by walking the instruction groups; see emitter::getLoopSize()).
 
                 if (block->bbJumpDest->isLoopAlign())
                 {
                     GetEmitter()->emitSetLoopBackEdge(block->bbJumpDest);
+
+                    if (block->bbNext != nullptr)
+                    {
+                        JITDUMP("Mark " FMT_BB " as label: alignment end-of-loop\n", block->bbNext->bbNum);
+                        block->bbNext->bbFlags |= BBF_HAS_LABEL;
```

Comment:
Shouldn't this be an assert that `block->bbNext->bbFlags` contains `BBF_HAS_LABEL` already? In what cases it won't have that flag? I see a comment in `genMarkLabelsForCodegen` that says `(fall-through flow doesn't require a label)`...was that always true or is that what we will start seeing with your change?

Reply:
No, this is a case where the following block will not necessarily have a label. It was always true that fall-through flow doesn't require a label, but before there were more cases where it "accidentally" did have a label. The alignment code never specified that it needed a label, so it would calculate loop sizes based on whatever happened to be the next label. After my change, we have fewer labels, so some loops ended up "too large" for alignment. Here, I force the label to exist when we know alignment wants one, for calculating loop sizes.

The one possibility is I could move this setting to `genMarkLabelsForCodegen`; I decided against it, but it doesn't really matter. Maybe centralizing the logic would be better.

Reply:
> so some loops ended up "too large" for alignment.

Yes, we should definitely compare loopsAligned  before and after to make sure we are aligning right amount of loops with this change.

Reply:
Over aspnet/benchmarks/libraries.crossgen/tests SPMI collections, I see loops/align candidates/loops aligned of 6135/1550/647 for baseline and the same bug 646 loops aligned for diff (my change). So it appears that exactly one fewer loop was aligned.

Reply:
Interesting. Did you figure out the CI failure and is it related?

Reply:
The assertion was due to the JIT deciding a loop was no longer a loop, then compacting the preheader with the former loop head, moving the align bit with the merge (with my new code). But since the (former) loop head no longer was the target of any branch, it didn't get a label. Solution: remove the align bit when this optimization happens.